#14020-Cart-Sales-Rule-with-negated-condition-over-special-price-does…

novikor opened the pull request:
…-not-work-for-configurable-products.

Use configurable product`s children for shopping cart rules validation
for cases when attribute exists for children only (E.g. special_price)

<!---
    Thank you for contributing to Magento.
    To help us process this pull request we recommend that you add the following information:
     - Summary of the pull request,
     - Issue(s) related to the changes made,
     - Manual testing scenarios,
-->

<!--- Please provide a general summary of the Pull Request in the Title above -->

### Description
<!---
    Please provide a description of the changes proposed in the pull request.
    Letting us know what has changed and why it needed changing will help us validate this pull request.
-->

### Fixed Issues (if relevant)
<!---
    If relevant, please provide a list of fixed issues in the format magento/magento2#<issue_number>.
    There could be 1 or more issues linked here and it will help us find some more information about the reasoning behind this change.
-->
1. magento/magento2#14020 Cart Sales Rule with negated condition over special_price does not work for configurable products

### Manual testing scenarios
<!---
    Please provide a set of unambiguous steps to test the proposed code change.
    Giving us manual testing scenarios will help with the processing and validation process.
-->
1. Enable the `special_price` for promo rule conditions: 
   1. Stores > Attributes > Product
   2. Search for and edit `special_price`
   3. Storefront Properties: Use for Promo Rule Conditions: Yes
   4. Save
2. create new cart sales rule:
   1. Marketing > Cart Price Rules
   2.  Add New Rule:
       1. Rule name = "Special Price Test"
       2. Website: select all
       3. Customer Groups: select all
       4. Coupon: "Specific Coupon"
       5. Coupon Code: "special"
       6. Actions: Discount Amount: "10"
       7. Actions: Apply the rule only to cart items matching the following conditions: "if **ALL** of these conditions are **FALSE**": "Special Price **greater than 0.01**"
       8. Leave all other fields to their defaults and save
3. In the front end, add a the child product of a configurable product to the cart, such that the child product has a special price set. This product should be the only one in the cart.
4. "View and Edit Cart" > "Apply Discount Code": `special`

### Contribution checklist
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] All automated tests passed successfully (all builds on Travis CI are green)

novikor committed:
#14020-Cart-Sales-Rule-with-negated-condition-over-special-price-does-not-work-for-configurable-products.
Use configurable product`s children for shopping cart rules validation
for cases when attribute exists for children only (E.g. special_price)

magento-cicd2 commented:
[![CLA assistant check](https://cla.dev.magento.com/pull/badge/not_signed)](https://cla.dev.magento.com/magento/magento2?pullRequest=16342) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.dev.magento.com/magento/magento2?pullRequest=16342) before we can accept your contribution.<br/>

magento-engcom-team commented:
Hi @novikor. Thank you for your contribution 
Here is some useful tips how you can test your changes using Magento test environment.
Add the comment under your pull request to deploy test or vanilla Magento instance:
- `@magento-engcom-team give me test instance` - deploy test instance based on PR changes
- `@magento-engcom-team give me {$VERSION} instance` - deploy vanilla Magento instance

For more details, please, review the [Magento Contributor Assistant](https://github.com/magento/magento2/wiki/Magento-Contributor-Assistant) documentation

phoenix128 reviewed (approved):
Hello @novikor ,
thank you for your contribution.
We appreciate your support.

phoenix128 committed:
Removed empty line due to failing static test

phoenix128 committed:
Removed dependency to configurable product

sidolov commented on app/code/Magento/SalesRule/Model/Rule/Condition/Product.php:
```diff
@@ -26,20 +28,47 @@ protected function _addSpecialAttributes(array &$attributes)
+     * @return \Magento\Catalog\Api\Data\ProductInterface|\Magento\Catalog\Model\Product
+     * @throws \Magento\Framework\Exception\NoSuchEntityException
+     */
+    protected function getProductToValidate(\Magento\Framework\Model\AbstractModel $model)
```
Please, make this method private

novikor replied:
Done, please review: fae98c01e29ea9ca47eca00cf2081ede1f729ccb & 9a35b459f1b31b2b8d40c6597bd90ac9c5c865fe

novikor committed:
#14020-Cart-Sales-Rule-with-negated-condition-over-special-price-does-not-work-for-configurable-products.

Revert "Removed dependency to configurable product"

This reverts commit d2a0de83c0f73dd850925b4a70ab2e5cf233d74b.

novikor committed:
#14020-Cart-Sales-Rule-with-negated-condition-over-special-price-does-not-work-for-configurable-products.
Made getProductToValidate method private according to Technical Guideline (2.7)

novikor committed:
#14020-Cart-Sales-Rule-with-negated-condition-over-special-price-does-not-work-for-configurable-products.
Added configurable product to dependencies

novikor commented:
@sidolov , are there any updates?

sidolov commented:
Hi @novikor , we are trying to avoid new dependencies across Magento modules and define another one it's not a good idea. If you have the specific case for the configurable product during the rule processing it would be a better use default Magento ways for extensibility (plugins, DI configuration, etc).

novikor committed:
Revert "#14020-Cart-Sales-Rule-with-negated-condition-over-special-price-does-not-work-for-configurable-products. Added configurable product to dependencies"

This reverts commit 887ee4a

novikor committed:
Merge branch '2.2-develop' into #14020-Cart-Sales-Rule-with-negated-condition-over-special-price-does-not-work-for-configurable-products

# Conflicts:
#	app/code/Magento/SalesRule/Model/Rule/Condition/Product.php

novikor committed:
#14020-Cart-Sales-Rule-with-negated-condition-over-special-price-does-not-work-for-configurable-products.
Restored \Magento\SalesRule\Model\Rule\Condition\Product to its previous state except PHPDocs.

novikor committed:
Moved adjustments to plugin.

novikor commented:
Done. Please review.

novikor committed:
#14020-Cart-Sales-Rule-with-negated-condition-over-special-price-does-not-work-for-configurable-products

Removed an empty line.

sidolov commented on app/code/Magento/ConfigurableProduct/Plugin/SalesRule/Model/Rule/Condition/Product.php:
```diff
@@ -0,0 +1,53 @@
+{
+    /**
+     * @param \Magento\SalesRule\Model\Rule\Condition\Product $subject
+     * @param \Magento\Framework\Model\AbstractModel          $model
```
Please, remove extra spaces

sidolov commented on app/code/Magento/ConfigurableProduct/Plugin/SalesRule/Model/Rule/Condition/Product.php:
```diff
@@ -0,0 +1,53 @@
+
+    /**
+     * @param \Magento\SalesRule\Model\Rule\Condition\Product $subject
+     * @param \Magento\Framework\Model\AbstractModel          $model
```
Please, remove extra spaces

sidolov commented on app/code/Magento/ConfigurableProduct/Plugin/SalesRule/Model/Rule/Condition/Product.php:
```diff
@@ -0,0 +1,53 @@
+<?php declare(strict_types=1);
```
Please, cover new code at least by the unit test. It would be great if scenario will be covered with integration test

novikor replied:
Unit tests coverage have been implemented.
@sidolov, please check.

novikor committed:
Merge remote-tracking branch 'upstream/2.2-develop' into #14020-Cart-Sales-Rule-with-negated-condition-over-special-price-does-not-work-for-configurable-products

novikor committed:
Merge remote-tracking branch 'origin/#14020-Cart-Sales-Rule-with-negated-condition-over-special-price-does-not-work-for-configurable-products' into #14020-Cart-Sales-Rule-with-negated-condition-over-special-price-does-not-work-for-configurable-products

novikor committed:
#14020-Cart-Sales-Rule-with-negated-condition-over-special-price-does-not-work-for-configurable-products.
Unit tests coverage: case when child should be used

novikor committed:
#14020-Cart-Sales-Rule-with-negated-condition-over-special-price-does-not-work-for-configurable-products.
Fixed mock objects comparison

novikor committed:
Covered case when simple product is being validated

novikor committed:
#14020-Cart-Sales-Rule-with-negated-condition-over-special-price-does-not-work-for-configurable-products.
Fixed static tests.

novikor committed:
Merge remote-tracking branch 'upstream/2.2-develop' into #14020-Cart-Sales-Rule-with-negated-condition-over-special-price-does-not-work-for-configurable-products

novikor commented:
@sidolov , any updates?

sidolov commented:
Hi @novikor , we found the problems with this solution on the B2B edition and trying to figure it out right now. 
Thank you for your patience.

magento-engcom-team commented:
Hi @novikor. Thank you for your contribution. 
We will aim to release these changes as part of 2.2.8. 
Please check the release notes for final confirmation.

Please, consider to port this solution to 2.3 release line.
You may use [Porting tool](https://devdocs.magento.com/guides/v2.2/contributor-guide/contributing.html#porting-tool) to port commits automatically.
